### PR TITLE
Provide before/after sign in callbacks for users 

### PR DESCRIPTION
**Pull request by allenwq**

### CODE REVIEW

File: lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module Extensions::SignInWithCallback::Devise::Controllers::SignInOut
+  class << self
+    def included(base)
```
Comment (lowjoel) [inline lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb]: I call this module_

File: lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      base.alias_method_chain :sign_in, :callback
+    end
+  end
+
+  def sign_in_with_callback(resource_or_scope, *args)
+    original_args = args.dup
+    original_args.extract_options!
+    resource = original_args.last || resource_or_scope
+
+    if resource.respond_to?(:_sign_in_callbacks)
```
Comment (lowjoel) [inline lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb]: I don't think that's a good way to check whether callbacks are supported.

File: spec/libraries/sign_in_with_callback_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    def after_callback
+    end
+  end
+
+  class self::TestDeviseController < ApplicationController
+    include Devise::Controllers::SignInOut
+  end
+
+  controller(self::TestDeviseController) do
+  end
```
Comment (lowjoel) [inline spec/libraries/sign_in_with_callback_spec.rb]: just include Devise::Controllers:SignInOut here.

File: spec/libraries/sign_in_with_callback_spec.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+RSpec.describe 'Extension: sign_in with resource', type: :controller do
```
Comment (lowjoel) [inline spec/libraries/sign_in_with_callback_spec.rb]: Devise: Signing in with resource

File: lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+      resource.run_callbacks :sign_in do
+        sign_in_without_callback(resource_or_scope, *args)
+      end
+    else
+      sign_in_without_callback(resource_or_scope, *args)
```
Comment (lowjoel) [inline lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb]: Add another test to run this line, to show that it works with models that do not have a sign_in callback.

File: lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb (hunk 2 above)
Comment (lowjoel) [inline lib/extensions/sign_in_with_callback/devise/controllers/sign_in_out.rb]: You can provide a module to include support for the sign_in callback. Then when signing in, check if that object has the module, if so, trigger the callback.

File: lib/autoload/sign_in_callbacks.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+module SignInCallbacks
```
Comment (lowjoel) [inline lib/autoload/sign_in_callbacks.rb]: You probably might want to namespace this under Devise, or User.

**Comment by allenwq** [comment]
@lowjoel all done, thanks !

File: app/models/concerns/user_authentication_concern.rb [hunk 7]
```diff
@@ -29,6 +32,10 @@ def email_required?
     true
   end
 
+  def create_instance_users
```
Comment (lowjoel) [inline app/models/concerns/user_authentication_concern.rb]: put this with the rest of the instance methods.

Comment (allenwq) [inline app/models/concerns/user_authentication_concern.rb]: Where are the reset of instance methods ?  I already put them under other instance methods in this concern.

Comment (lowjoel) [inline app/models/concerns/user_authentication_concern.rb]: ah, i didnt see the break, I thought it was after include

**Pull request merged**